rep-2000.rst

```diff
@@ -547,6 +547,124 @@ Build System Support:
+|    arm32     |     Tier 2 [a][s]     |                      |                      |     Tier 3 [s]     |   Tier 3 [s]  |
++--------------+-----------------------+----------------------+----------------------+--------------------+---------------+
+
+The following indicators show what delivery mechanisms are available for each platform.
+
+" [d] " Debian packages will be provided for this platform for packages submitted to the rosdistro.
+
+" [a] " Binary releases are provided as a single archive per platform containing all packages in the Eloquent ROS 2 repos file [6]_.
+
+" [s] " Compilation from source.
+
+Middleware Implementation Support:
+
++--------------------------+---------------------+---------------+-----------------------------+--------------------------------------+
+| Middleware Library       | Middleware Provider | Support Level | Platforms                   | Architectures                        |
++==========================+=====================+===============+=============================+======================================+
+|  rmw_fastrtps_cpp        | eProsima Fast-RTPS  | Tier 1        | All Platforms               | All Architectures                    |
++--------------------------+---------------------+---------------+-----------------------------+--------------------------------------+
+|  rmw_connext_cpp         | RTI Connext         | Tier 1        | All Platforms except Debian | All Architectures except arm64/arm32 |
+|                          |                     |               | and OpenEmbedded            |                                      |
++--------------------------+---------------------+---------------+-----------------------------+--------------------------------------+
+| rmw_cyclonedds_cpp       | Eclipse Cyclone DDS | Tier 2        | All Platforms               | All Architectures                    |
++--------------------------+---------------------+---------------+-----------------------------+--------------------------------------+
+| rmw_opensplice_cpp       | ADLINK OpenSplice   | Tier 2        | All Platforms except Debian | All Architectures                    |
+|                          |                     |               | and OpenEmbedded            |                                      |
++--------------------------+---------------------+---------------+-----------------------------+--------------------------------------+
+| rmw_fastrtps_dynamic_cpp | eProsima Fast-RTPS  | Tier 2        | All Platforms               | All Architectures                    |
++--------------------------+---------------------+---------------+-----------------------------+--------------------------------------+
+
+Middleware implementation support is dependent upon the platform support tier.
+For example a Tier 1 middleware implementation on a Tier 2 platform can only receive Tier 2 support.
+
+Minimum language requirements:
+
+- C++17
```

Comment:
I believe there are some users which cannot easily update to C++17 due to limited access to newer compiler versions.
Unless we get an OK from people who have been limited to C++14 in the past, I don't think it's wise to update the minimum requirement to C++17 just for new language features.

@dejanpan FYI

Reply:
Please share more precise feedback, e.g. what exact platform / environment / compiler version are you referring to.

Reply:
I don't know if I can share it, that's why I pinged @dejanpan.

I do know that we've discussed this in the past and it was an issue though. I don't see why it would be an issue now either.

Reply:
@dejanpan Friendly ping.

Reply:
I don't have precise feedback (and personally would love to move to C++17), but in general compilers for the embedded world are a geological age behind.

Reply:
> in general compilers for the embedded world are a geological age behind.

We are hearing this generic statement a lot. I would like to get concrete examples what people are actually using to make an informed decision.

Reply:
I agree, we need to know names of compilers and how much they are actually being used in ROS. Otherwise it's an empty argument, unfortunately, and we would end up going with the commonly-used compilers on the major platforms.

Reply:
> in general compilers for the embedded world are a geological age behind.

Reply:
> in general compilers for the embedded world are a geological age behind.

This isn't the case for OpenEmbedded: in the release series that `meta-ros` currently supports (**thud**), gcc v8.2.0 is provided (which is later than the version provided by Ubuntu **bionic**) -- see [Package Version Differences](https://github.com/ros/meta-ros/wiki/Package-Version-Differences).

Reply:
@dirk-thomas @gbiggs @herb-kuta-lge I will give a longer answer that will hopefully also be educational as per https://github.com/ros-infrastructure/rep/pull/217/files#r367716644.

What we do at Apex.AI is fork C and C++ core packages from https://github.com/ros2/, modify, extend and certify them and then sell licenses with the professional support. 

For this discussion relevant are these steps that happen between the forking and the deployment to the customers:
1. Code is compiled with additional security and safety related compiler flags (e.g. `-Werror=format-security`, `-Wcast-align`, ...)
1. Code is analysed with the [PCLint](https://gimpel.com/?gclid=EAIaIQobChMIo4SamoCd5wIVRhitBh0PGQhTEAAYASAAEgI49fD_BwE) static code analyser using [Adaptive Autosar Coding Guidelines](https://www.autosar.org/fileadmin/user_upload/standards/adaptive/19-11/AUTOSAR_RS_CPP14Guidelines.pdf)
1. Code is analysed with the [Vector Cast](https://www.vector.com/us/en-us/products/products-a-z/software/vectorcast/?gclid=EAIaIQobChMIzYmL1oCd5wIVC8RkCh1GkQT8EAAYASAAEgKeV_D_BwE#c74806) tool and the code coverage is computed for lines, functions, branches and [MC/DC](https://ntrs.nasa.gov/archive/nasa/casi.ntrs.nasa.gov/20010057789.pdf)
    1. Code coverage is brought to >90%
1. Architecture and the dependency structure of the code is analysed with [Lattix Architect](https://www.lattix.com/products-architecture-issues/)
1. Test stubs are (auto) created with VectorCast for the dependent units (system calls, ...)
1. The code is sanitized with the various sanitizers: https://clang.llvm.org/docs/index.html
1. The code is cross-compiled for Embedded Linux and QNX on ARM ECUS
    1. In the near-term future other POSIX-compliant [RTOSes](https://en.wikibooks.org/wiki/Embedded_Systems/Common_RTOS) will be supported
1. We use the gcc and libcxx versions available on above RTOSes and which have been validated with the validation suites like: https://solidsands.com/supertest.
1. Following back-ported features from C++17 are used: std::optional, std::variant, std::string_view, std::apply, polymorphic allocator
1. rmw layer for [RTI Connext DDs Micro](https://www.rti.com/products/connext-dds-micro) is created
1. The code undergoes performance testing using https://gitlab.com/ApexAI/performance_test
1. The performance is analysed and tuned with the following tools: perf, heaptrack, valgrind, LTT-ng


For us to now decide whether we can switch to the newest version of C++ we need to consider if above RTOSes and tools support, in this case, C++17.

---
For long story short, I will below list only the things that we currently do not know in terms of switching from C++14 to C++17:
1. QNX has not publicly acknowledged support for C++17. In fact from above listed RTOSes, VxWorks is the only one so far that listed [C++17 support](https://www.windriver.com/news/press/pr.html?ID=22444)
1. Adaptive Autosar Coding Guidelines for C++17 are not available yet
1. Someone will have to create a toolchain file for the cross compilation on the ARM-based ECUs

In addition we do not know what is the willingness of the (automotive) customers to switch to C++17. So far everyone is set on C++14.

In conclusion: **I would like to make my proposal based on the plans of https://www.autosar.org/standards/adaptive-platform/. If you can, allow me couple of days to track down the right person for that.**

Reply:
> I would like to make my proposal based on the plans of https://www.autosar.org/standards/adaptive-platform/. If you can, allow me couple of days to track down the right person for that.

@dejanpan Do you have any update on this? I would like to finalize this topic within this week.

Reply:
We were thinking that polymorphic allocators could address a problem when using custom allocators for messages (either in the case RMW impl. allocate the message internally which doesn't match the allocator the user code uses or when passing messages for intra-process comms). After investigating these cases we discovered that atm you can't use the `rclcpp` API to create e.g. publisher / subscribers for messages using a custom / non-default allocator. As a consequence a mismatch of allocators simply can't occur.

While this (using messages with custom allocators in the rclcpp API) should be resolved it is unlikely that we will address the problem for Foxy (with only a week to the API freeze date) since enabling this use case will require some significant refactoring of the C++ typesupport API. As a consequence we don't have a need to use polymorphic allocators.

Due to the brought up concerns we will stick to C++14 as the minimum C++ standard in Foxy: see 953a64d7a351c965198e0d483b19af54e30b6a52.